NDTensors/src/lib/SymmetrySectors/src/abstractsector.jl

```diff
@@ -0,0 +1,114 @@
+# This file defines the abstract type AbstractSector
+# all fusion categories (Z{2}, SU2, Ising...) are subtypes of AbstractSector
+
+using BlockArrays: blocklengths
+using ..LabelledNumbers: LabelledInteger, label, label_type, labelled, unlabel, unlabel_type
+using ..GradedAxes: GradedAxes, blocklabels, fuse_blocklengths, gradedrange, tensor_product
+
+abstract type AbstractSector end
+
+# ===================================  Base interface  =====================================
+function Base.isless(c1::C, c2::C) where {C<:AbstractSector}
+  return isless(sector_label(c1), sector_label(c2))
+end
+
+# =================================  Sectors interface  ====================================
+trivial(x) = trivial(typeof(x))
+function trivial(axis_type::Type{<:AbstractUnitRange})
+  return gradedrange([trivial(eltype(axis_type))])  # always returns nondual
+end
+function trivial(la_type::Type{<:LabelledInteger})
+  return labelled(one(unlabel_type(la_type)), trivial(label_type(la_type)))
+end
+function trivial(type::Type)
+  return error("`trivial` not defined for type $(type).")
+end
+
+istrivial(c::AbstractSector) = (c == trivial(c))
+
+function sector_label(c::AbstractSector)
+  return error("method `sector_label` not defined for type $(typeof(c))")
+end
+
+block_dimensions(g::AbstractUnitRange) = block_dimensions(SymmetryStyle(g), g)
+block_dimensions(::AbelianStyle, g) = unlabel.(blocklengths(g))
+function block_dimensions(::NotAbelianStyle, g)
+  return quantum_dimension.(blocklabels(g)) .* blocklengths(g)
+end
+
+quantum_dimension(x) = quantum_dimension(SymmetryStyle(x), x)
+
+function quantum_dimension(::NotAbelianStyle, c::AbstractSector)
+  return error("method `quantum_dimension` not defined for type $(typeof(c))")
+end
+
+quantum_dimension(::AbelianStyle, ::AbstractSector) = 1
+quantum_dimension(::AbelianStyle, g::AbstractUnitRange) = length(g)
+quantum_dimension(::NotAbelianStyle, g::AbstractUnitRange) = sum(block_dimensions(g))
+
+# ===============================  Fusion rule interface  ==================================
+⊗(c1::AbstractSector, c2::AbstractSector) = fusion_rule(c1, c2)
+
+function fusion_rule(c1::AbstractSector, c2::AbstractSector)
+  return fusion_rule(combine_styles(SymmetryStyle(c1), SymmetryStyle(c2)), c1, c2)
+end
+
+function fusion_rule(::NotAbelianStyle, c1::C, c2::C) where {C<:AbstractSector}
+  degen, sectors = label_fusion_rule(C, sector_label(c1), sector_label(c2))
```

Comment:
A recommendation for the format of `label_fusion_rule` would be to output a `Vector{<:Pair{<:AbstractSector,Int}}`, i.e. output a single vector that is a pair of the irreps/sectors and their corresponding block length/degeneracy.

Outputting two vectors of the same length doesn't feel quite right to me, and also I think it makes more sense to list the irreps/sectors first since that is the more "important" information, in my mind. In other words, the degeneracy/block length is secondary information about the irrep/sector, so should be listed second.

There are a few precedents for that format, like Julia's `Dict` constructor, ITensor's current `QN`/`Index` constructor, and also one of the `gradedrange` constructors:
```julia
julia> Dict(["x" => 2, "y" => 3])
Dict{String, Int64} with 2 entries:
  "x" => 2
  "y" => 3

julia> using ITensors: Index, QN

julia> Index([QN("N", 0) => 2, QN("N", 1) => 3])
(dim=5|id=968) <Out>
 1: QN("N",0) => 2
 2: QN("N",1) => 3

julia> using NDTensors.GradedAxes: gradedrange

julia> using NDTensors.Sectors: U1

julia> gradedrange([U1(0) => 2, U1(1) => 3])
2-blocked 5-element BlockArrays.BlockedOneTo{NDTensors.LabelledNumbers.LabelledInteger{Int64, U1}, Vector{NDTensors.LabelledNumbers.LabelledInteger{Int64, U1}}}:
 NDTensors.LabelledNumbers.LabelledInteger{Int64, U1}(1, U1(0))
 NDTensors.LabelledNumbers.LabelledInteger{Int64, U1}(2, U1(0))
 ──────────────────────────────────────────────────────────────
 NDTensors.LabelledNumbers.LabelledInteger{Int64, U1}(3, U1(1))
 NDTensors.LabelledNumbers.LabelledInteger{Int64, U1}(4, U1(1))
 NDTensors.LabelledNumbers.LabelledInteger{Int64, U1}(5, U1(1))
```